src/outpack/location_ssh.py

```diff
@@ -14,43 +13,48 @@
 from outpack.metadata import MetadataCore, PacketFile, PacketLocation
 from outpack.static import LOCATION_LOCAL
 
-SCP_URL = re.compile(r"((?P<username>[^@]+)@)?(?P<hostname>[^:]+):(?P<path>.*)")
-
 
 def parse_ssh_url(url):
     """
     Parse the URL of an SSH location.
 
-    We follow Git's definition of URLs, and support two syntaxes: a short
-    `username@hostname:path` version, inspired by the scp command, and a longer
-    more explicit `ssh://username@hostname:port/path`.
+    URLs follow the form `ssh://username@hostname:port/path`, where path is
+    interpreted as an absolute path (ie. the separating slash at the start of
+    the path is included in it).
+
+    However, if the path begins with `/~/` (as in `ssh://hostname/~/foo/bar`),
+    then the rest of the path is assumed to be relative to the user's home
+    directory.
 
-    The SCP syntax doesn't support configuring a port number (though an
-    .ssh/config file can be used to work around that limitation).
+    This generally follows Git's semantics. Git also supports more complicated
+    expansions (eg. using another user's home directory, such as `/~alice/`),
+    but it implements this by having a server-side binary, which we don't have.
 
-    By default, the path in the SCP syntax is relative to the remote user's home
-    directory, and `hostname:/foo/bar` can be used to specify an absolute path.
-    The explicit URL syntax is always absolute.
+    Git also supports an alternative SCP-like syntax, of the form
+    `username@hostname:path`, where `path` is relative by default, and port
+    numbers cannot be specified. The lack of scheme to identify the protocol
+    makes these URLs inconvenient, so we don't support it.
 
     See https://git-scm.com/docs/git-pull#_git_urls.
     """
-    if "://" in url:
-        parts = urlparse(url)
-        if parts.scheme != "ssh":
-            msg = f"Protocol of SSH url must be 'ssh', not '{parts.scheme}'"
-            raise Exception(msg)
+    parts = urlparse(url)
+    if parts.scheme != "ssh":
+        msg = "Protocol of SSH url must be 'ssh'"
+        raise Exception(msg)
 
-        if parts.path == "":
-            msg = "No path specified for SSH location"
-            raise Exception(msg)
+    if parts.path == "":
+        msg = "No path specified for SSH location"
+        raise Exception(msg)
 
-        return parts.username, parts.hostname, parts.port, parts.path
+    path = PurePosixPath(parts.path)
 
-    elif m := SCP_URL.fullmatch(url):
-        return (m.group("username"), m.group("hostname"), None, m.group("path"))
-    else:
-        msg = f"Invalid SSH url: '{url}'"
-        raise Exception(msg)
+    # This condition is equivalent to path.is_relative_to("/~"), except that
+    # function isn't available on Python 3.8.
+    home = PurePosixPath("/~")
```

Comment:
I think I prefer the mercurial approach here tbh. How strongly do you feel about this?

My reasoning is that relative paths will probably be more commonly used (most people will be used to `scp ...` popping files in their home directory unless they take additional action. I know that this is slightly inconsistent with http urls though.

Reply:
Yeah I think I actually prefer the mercurial approach too. My main reason for doing it this way is that more people will be familiar with Git than Mercurial.

That being said, probably very few people actually use Git in this way and instead use whatever magic GitHub does, so maybe it doesn’t matter. I’ll change it to the mercurial semantics.